python/pyarrow/types.pxi

```diff
@@ -357,6 +357,57 @@ cdef class DataType(_Weakrefable):
                                            _as_c_pointer(in_ptr)))
         return pyarrow_wrap_data_type(result)
 
+    def __arrow_c_schema__(self):
+        """
+        Export to a ArrowSchema PyCapsule
+
+        Unlike _export_to_c, this will not leak memory if the capsule is not used.
+        """
+        cdef:
+            ArrowSchema* c_schema = <ArrowSchema*>malloc(sizeof(ArrowSchema))
+        with nogil:
+            check_status(ExportType(deref(self.type), c_schema))
+        return cpython.PyCapsule_New(c_schema, 'arrowschema', &pycapsule_schema_deleter)
+
+    @staticmethod
+    def _import_from_c_capsule(schema):
```

Comment:
I think it would also be nice to call this automatically from `pa.schema`, what do you think?